### PR TITLE
Issue 622 - Check performance of hypergeometric distribution when number of infections to draw is larger than population

### PR DESCRIPTION
**Pull request by FiodorG**

### CODE REVIEW

**Comment by FiodorG** [comment]
There are no "logical" and natural ways to fix this I could think about. We are given a unique contact number from the mixing matrix through time (*multipliers), and that number is an average over all people. So there is nothing "naturally" preventing the contacts to be smaller than total population in node. Given this, we have two choices:
1) Simple cap the contacts at total population like I have done. This implies the cap has a higher probability of happening
2) Resample Poissons until all of them are below the cap. This is called a truncated Poisson distribution. Or sample many more than required and select enough that satisfy our criteria. I don't like this solution because it requires some iterative calculation. However this is technically more correct. How to do it is described here: https://stackoverflow.com/questions/52398903/scipy-poisson-distribution-with-an-upper-limit 

I believe given it will again happen on low population nodes, it's not so impactful that we should add an iterative Poisson sampling, and I propose to go ahead with 1).


Example to illustrate the above. Assume we sample from a uniform distribution in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], each with probability 0.1.
1) Cap at 5. This means we use the value of 5 for any number >= 5. So 5 will be sampled 50% of the time
2) Resample if >5. This means we effectively sample in [1, 2, 3, 4, 5] with probas 0.2 each.

File: simple_network_sim/network_of_populations/__init__.py [hunk 1]
```diff
@@ -584,6 +583,8 @@ def _computeInfectiousContactsStochastic(
     numberOfContacts = random_state.poisson(contacts, size=int(infectious))
```
Comment (aflag) [inline simple_network_sim/network_of_populations/__init__.py:583]: Shouldn't we cap the number of contacts before calling poisson?

File: simple_network_sim/network_of_populations/__init__.py [hunk 2]
```diff
@@ -584,6 +583,8 @@ def _computeInfectiousContactsStochastic(
...
     # If 0, the hypergeometric calls fails, but mathematically is 0
     numberOfContacts = numberOfContacts[numberOfContacts > 0]
+    # Number of contacts cannot be higher than the number of people in the node
+    numberOfContacts = np.minimum(numberOfContacts, int(totalInAge))
```
Comment (aflag) [inline simple_network_sim/network_of_populations/__init__.py:587]: Actually, why can't the number of contacts be bigger than the population? One person can contact more than one other people. I think the actual limit is `totalInAge**2`, isn't it?

File: simple_network_sim/network_of_populations/__init__.py (hunk 1 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:583]: We could do as well, that would lower the samples coming out of Poisson. But we'd still need to cap the outputs of Poisson. as it's just the average the Poisson will be centered around.

File: simple_network_sim/network_of_populations/__init__.py (hunk 2 above)
Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:587]: I think as per the citation you found in a previous PR, the contacts in the mixing matrix are "unique people in a certain age group that you have been in contact with". So each infected person had on average n contacts with people in age group. So if there are 100 people in some age group, and you are a person, the max number of contacts you can report is 100-1.

So technically I should cap it at `int(totalInAge)-1`

Comment (FiodorG) [reply simple_network_sim/network_of_populations/__init__.py:587]: Your comment would be correct if we summed up all the contacts before passing in the hypergeom. But the `numberOfContacts` is an array of contacts per each person. Hence the `n-1` cap. 

Does that make sense?

File: simple_network_sim/network_of_populations/__init__.py (hunk 1 above)
Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:583]: Ah yeah, I see the problem now.

File: simple_network_sim/network_of_populations/__init__.py (hunk 2 above)
Comment (aflag) [reply simple_network_sim/network_of_populations/__init__.py:587]: Ah, yeah, that makes sense.

**Pull request merged**